proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+        public ActivityTraceId TraceId { get; }
+        public ActivitySpanId SpanId { get; }
+        public ActivityTraceFlags TraceFlags { get; }
+        public string? TraceState { get; }
+
+        public static bool operator ==(ActivityContext left, ActivityContext right);
+        public static bool operator !=(ActivityContext left, ActivityContext right) ;
+        public bool Equals(ActivityContext value);
+        public override bool Equals(object? obj);
+        public override int GetHashCode();
+    }
+}
+```
+
+### ActivityEvent
+
+```C#
+namespace System.Diagnostics
+{
+    /// A text annotation associated with a collection of attributes.
+    /// Event has name and optionally has timestamps and attributes
+    public readonly struct ActivityEvent
+    {
+        /// Initializes a new instance of the <see cref="ActivityEvent"/> class.
+        public ActivityEvent(string name) : this(name, DateTimeOffset.UtcNow, s_emptyAttributes);
+        public ActivityEvent(string name, DateTimeOffset timestamp) : this(name, timestamp, s_emptyAttributes);
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes);
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes);
+
+        public string Name { get; }
+        public DateTimeOffset Timestamp { get; }
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+    }
+}
+```
+
+ ### ActivityLink
+
+```C#
+namespace System.Diagnostics
+{
+    /// Activity may be linked to zero or more other <see cref="ActivityContext"/> that are causally related.
+    /// Links can point to ActivityContexts inside a single Trace or across different Traces.
+    /// Links can be used to represent batched operations where a Activity was initiated by multiple initiating Activities,
+    /// each representing a single incoming item being processed in the batch.
+    public readonly struct ActivityLink : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes = null);
+        public ActivityContext Context { get; }
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+        
+        public override bool Equals(object? obj);
+        public bool Equals(ActivityLink value);
+        public static bool operator ==(ActivityLink left, ActivityLink right);
+        public static bool operator !=(ActivityLink left, ActivityLink right);
+        public override int GetHashCode();
+    }
+}
+```
+
+### ActivityDataRequest
+
+```C#
+namespace System.Diagnostics
+{
+    /// Used by ActivityListener to indicate what amount of data should be collected for this Activity
+    /// Requesting more data causes greater performance overhead to collect it.
+    public enum ActivityDataRequest
```

Comment:
Whose responsibility is to act based on `ActivityDataRequest`? If I'm the library author, where do I use `ActivityDataRequest` in the following sample?
```
using (Activity? activity = source.StartActivity("Request"))
    {
        activity?.AddTag..
        // How do I know how much data to collect here? 
    }
```

Reply:
It looks to me like the Library author will get a binary information only - i.e whether there is a listener or not for the Activity.
or does library author gets `ActivityDataRequest` itself, so authors can decide to emit more info based on value for `ActivityDataRequest`.

Reply:
ActivityDataRequest is useful only when there is a listener and the Library authors will not use ActivityDataRequest. Library author shouldn't care about anything more than if the Activity object is created and in which state (propagation only, or collecting data or intended to be recorded). If there are no listeners, StartActivity will just return null and no Activity object will get created. If there are listeners, the listener's callbacks will return ActivityDataRequest to help tell if we need to create the Activity and with what state.

Reply:
Okay. But how does library authors know the state of activity object to determine if it is propogation only, collecting data or intended to be recorded?
I can see Activity having a method `IsAllDataRequested` returning bool. Are there similar methods like `IsPropogationOnly` etc on Activity?

Reply:
When Activity.IsAllDataRequested = false means it is for propagation only.